### PR TITLE
Dev with pdks

### PR DESCRIPTION
**Pull request by dan-fritchman**

### CODE REVIEW

**Comment by dan-fritchman** [comment]
@ThomasPluck was #111 passing the CI suite before I merged it?

**Comment by ThomasPluck** [comment]
Good morning. No, the tests aren't working yet. My guess is that it's the test container again.

**Comment by ThomasPluck** [comment]
Guess we're at the "idk wtf is going on here" stage of debugging, now. I also threw in the digital cells as a freebie #119.

**Comment by dan-fritchman** [comment]
Especially with this whole digital cell library, this has gotten pretty huge.  
There is presumably some other scripting/ code behind collecting those ~14k lines worth of standard cells right? (I don't imagine you wrote that from scratch!) 

Would it make sense to: 

* (a) Share/ publish that code, and/ or
* (b) Keep its output as a separately tracked file

This feels a lot like doing a huge binding-generation to a big C library. Hopefully it never has to happen again? But realistically, it will, someone will find something some day. 

Also - I'm gonna make the core `hdl21/` parts of this a separate PR.

**Comment by dan-fritchman** [comment]
Core `hdl21/` contents are merged via #121.  
And any commentary about CI failing because my account hit some quotas - well, wrong, forget that.  
https://github.com/dan-fritchman/Hdl21/actions/runs/5060811696 for example is passing.  
Seems the container setup is where the problem lies.

**Comment by ThomasPluck** [comment]
> There is presumably some other scripting/ code behind collecting those ~14k lines worth of standard cells right?

Just a simple script that iterates through standard cell libraries and looks for `.subckt` instances, I can add this as `sky130_sc_build.py` if you like. The extra pdk_logic is pretty simple and just wraps everything up so that logic circuits work as expected.

> Seems the container setup is where the problem lies.

That indeed seems to be the case, it is difficult to test this stuff locally. I'll try to resolve this cryptic issue ASAP. VLSIR might need a new version pushed looking at the failure pattern on #121

**Comment by dan-fritchman** [comment]
Yeah, I think it would be good to share that script. Maybe in like a `Sky130/scripts/` directory?  
Ideally we can store its output as a single file, and any "massaging" can exist outside that file. 

As for the container - an alternative to support `ngspice` only is to just use GitHub's default ubuntu, and `apt install` its  "unstable", "testing", "experimental", or whatever non-decade-old edition gets published as a package. 

Xyce would be a different story, since it only gets binary-distributed for RedHat-like Linuxes.

**Comment by ThomasPluck** [comment]
The new `vlsir-test-container` has Tim Edward's Open PDKs as well as Xyce installed. It'd be nice to be getting the spice simulations running and not just throwing an error in CI. Unless this is a minimal concern and we just want to merge Sky130 ASAP.

**Comment by dan-fritchman** [comment]
I think given where they stand, the right approach is probably to split into two PRs. 

* (1) with the simulation CI setup
* (2) with the PDK additions 

And (2) may need to at least temporarily skip some tests, if (1) isn't ready.

**Comment by ThomasPluck** [comment]
Okay, so tests are now working as intended, the problem that is failing is that a new tag for `VLSIR` is needed to work with all of the new features included in the PDK.

**Comment by dan-fritchman** [comment]
Great! 
I think if we merge the current head of `main` into this, it should(?) (at least come close to?) work!

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -23,24 +24,31 @@ jobs:
...
-          python-version: ${{ matrix.python-version }}
+        run: |
+          apt-get install -y python${{ matrix.python-version }}
+          apt-get install -y python${{ matrix.python-version }}-dev
+          apt-get install -y python${{ matrix.python-version }}-venv
+          apt-get install -y python${{ matrix.python-version }}-distutils
 
-      - name: Install dependencies
-        run: bash scripts/install-${{ matrix.dep-installer }}.sh
+      #! The bizarre insanity of ". ./venv/..." is an Ubuntu sh translation of
```
Comment (dan-fritchman) [inline .github/workflows/test.yaml:39]: Do we really need all this? 
These tests run in an isolated container; does it still need a Python `venv` inside of there? The system `pythonX` won't do?

File: .github/workflows/test.yaml [hunk 2]
```diff
@@ -23,24 +24,31 @@ jobs:
...
+      #! "source venv/..." seen in most Python documentation
 
-      - name: Test with pytest
-        run: |
-          pytest -sv --cov=./ --cov-report=xml
+      - name: update `pip`
+        run: python${{ matrix.python-version }} -m pip install --upgrade pip
 
+      - name: pytest `HDL21`
```
Comment (dan-fritchman) [inline .github/workflows/test.yaml:45]: We also want to keep the "test with either PyPi or GitHub versions of VLSIR" option that was removed from here.

**Comment by dan-fritchman** [comment]
I took the liberty of merging recent `main` into here.  
For me, locally, it fails in the veins of #44 and #124.  
The CI/ test container is another matter; added a few comments on the CI config.

File: .github/workflows/test.yaml (hunk 1 above)
Comment (ThomasPluck) [reply .github/workflows/test.yaml:39]: Good point, let's check it out.

**Comment by ThomasPluck** [comment]
Everything is working for me locally, this is a versioning issue that arises with `pip install -e ".[dev]"` call in `scripts/install_dev.sh`. I can either fix this by introducing a 3rd script which ignores versioning requirements of HDL21, or I can wait for `Vlsir` to have a `4.0.1` version that stops HDL21 triggered pip to remove the local dev version installed.

**Comment by dan-fritchman** [comment]
So the trick for the VLSIR GitHub dependency was to revert `Hdl21/setup.py`'s requirement back from `4.0.1`, which doesn't exist, anywhere, to `4.0.dev0`, which at least exists in the GitHub-HEAD of the VLSIR tree. 

Perhaps there are better schemes by which these projects can sync version requirements. I'm all ears if so. But the primary one in use is "they just all have the same version string all the time". 

Again I took the liberty of pushing a few commits, which get it closer, but still failing to find stuff around that container.

**Comment by ThomasPluck** [comment]
Ready to merge, barring code coverage/VLSIR pypi issues.

**Comment by dan-fritchman** [comment]
Ship it!

**Pull request merged**